snuba/query/expressions.py

```diff
@@ -0,0 +1,165 @@
+    ) -> Iterator[Expression]:
+        """
+        Traverses the children of a hierarchical container like a tree.
+        """
+        for child in children:
+            if isinstance(child, ExpressionContainer):
+                for sub in child:
+                    yield sub
+            else:
+                yield child
+
+    def _transform_children(self,
+        children: Sequence[Expression],
+        func: Callable[[Expression], Expression],
+    ) -> Sequence[Expression]:
+        """
+        Transforms in place the children of a hierarchical node by applying
+        a mapping function.
+        """
+        def process_child(param: Expression) -> Expression:
+            r = func(param)
+            if r == param and isinstance(r, ExpressionContainer):
+                # The expression was not replaced by the function, which
+                # means it was unchanged. This means we need to traverse
+                # its children.
+                r.transform(func)
+            return r
+
+        return list(map(process_child, children))
+
+
+@dataclass
+class AliasedExpression(Expression, ExpressionContainer):
+    """
+    Wraps an expression and provides it an alias in the query.
+    This is an expression itself, thus it will appear when iterating
+    over the query.
+    """
+    alias: Optional[str]
+    node: Expression
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> None:
+        """
+        Applies the transformation to the aliased node.
+        """
+        self.node = self._transform_children((self.node,), func)[0]
+
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Traverses the wrapped node.
+        """
+        yield self
+        for e in self._iterate_over_children([self.node]):
+            yield e
+
+
+class Null(Expression):
+    def format(self) -> str:
+        raise NotImplementedError
+
+
+@dataclass
+class Column(Expression):
+    """
+    Represent a column in the schema of the dataset.
+    """
+    column_name: str
+    table_name: Optional[str]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+
+@dataclass
+class FunctionCall(Expression, ExpressionContainer):
+    """
+    Represents an expression that resolves to a function call on Clickhouse
+    """
+    function_name: str
+    parameters: Sequence[Expression]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> None:
+        """
+        The children of a FunctionCall are the parameters of the function.
+        Thus map runs func on the parameters, not on the function itself.
+        """
+        self.parameters = self._transform_children(self.parameters, func)
+
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Traverse the subtree in a prefix order.
+        """
+        yield self
```

Comment:
yes, because iterating over a function would never yield the function see the result of the iteration in the test cases.
> test_expressions::test_iterate